### PR TITLE
Feature: Requirements config option and handling

### PR DESCRIPTION
**Pull request by screamerbg**
This PR tunes how Mbed OS python requirements are being installed:
* Add support for (global & local) config flag 'no_requirements', which disables the built in check for installed python modules.
* Mbed CLI should attempt to auto-install requirements when:
   * Codebased is being imported (`mbed import`, `mbed add`, `mbed deploy`, `mbed new`)
   * Codebased is being updated/checked out (`mbed update`)
   * Certificates are being installed or changed (`mbed dm`)
* Mbed CLI should only check and NOT try to auto-install requirements when:
   * Code is being compiled, tested or exported
   * Any Mbed CLI operation that doesn't imply network connectivity

As thoroughly described here https://github.com/ARMmbed/mbed-cli/issues/756.

### CODE REVIEW

**Comment by screamerbg** [comment]
Related to #795, #782 and #756

File: mbed/mbed.py [hunk 1]
```diff
@@ -1618,7 +1618,12 @@ def requirements_contains(self, library_name):
         with open(os.path.join(req_path, req_file), 'r') as f:
             return library_name in f.read()
 
-    def check_requirements(self, show_warning=False):
+    def check_requirements(self, show_warning_only=False):
+        skip_requirements = self.get_cfg("NO_REQUIREMENTS", False)
+        if skip_requirements:
+            action("Skipping installed requirements check due to configuration flag.")
```
Comment (theotherjimmy) [inline mbed/mbed.py]: Could we make this a verbose only print?

File: setup.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
 setup(
     name="mbed-cli",
-    version="1.8.3",
+    version="1.8.4",
```
Comment (theotherjimmy) [inline setup.py]: It this supposed to be here?

Comment (screamerbg) [reply setup.py]: Yes. Read the description. This was supposed to get in first for a quick 1.8.4 release to hotfix the issue. Due to the new feature now merged in, this will have to become 1.9.0

File: mbed/mbed.py (hunk 1 above)
Comment (screamerbg) [reply mbed/mbed.py]: This is a non standard behavior and it's very easy to forget it. An extra line is not an overkill and it passes the responsibility to the user to handle requirements. Without the notice, there's no indication that one of the fundamental mbed CLI functions is not being performed because of user config.

Comment (bridadan) [reply mbed/mbed.py]: I think having it under a verbose print makes sense, we already tell the user to run with `-v` if the subprocess exits with a non zero code: https://github.com/ARMmbed/mbed-cli/blob/master/mbed/mbed.py#L3388

File: mbed/mbed.py [hunk 3]
```diff
@@ -53,7 +53,7 @@
 
 
 # Application version
-ver = '1.8.3'
+ver = '1.8.4'
```
Comment (bridadan) [inline mbed/mbed.py]: We should put this in a separate PR when we're ready to make a release.

File: mbed/mbed.py [hunk 4]
```diff
@@ -2977,7 +2982,7 @@ def test_(toolchain=None, target=None, compile_list=False, run_list=False,
 def dev_mgmt(toolchain=None, target=None, source=False, profile=False, build=False):
     orig_path = getcwd()
     program = Program(getcwd(), True)
-    program.check_requirements(True)
+    program.check_requirements()
```
Comment (bridadan) [inline mbed/mbed.py]: This is a behavior change, this will now cause the CLI to halt and exit with a non-zero return code if the requirements aren't present before running the `mbed dm` command.

I'm not necessarily against this, it just wasn't documented in the PR description or the commit message so I missed this at first.

File: mbed/mbed.py [hunk 5]
```diff
@@ -1681,7 +1686,7 @@ def post_action(self, check_reqs=True):
             shutil.copy(os.path.join(mbed_tools_path, 'default_settings.py'), os.path.join(self.path, 'mbed_settings.py'))
 
         if check_reqs:
-            self.check_requirements(True)
+            self.check_requirements()
```
Comment (bridadan) [inline mbed/mbed.py]: This is a behavior change, this will now cause the CLI to halt and exit with a non-zero return code if the requirements aren't present after cloning (and auto-installing) mbed-os.

I'm not necessarily against this, it just wasn't documented in the PR description or the commit message so I missed this at first.

File: mbed/mbed.py [hunk 6]
```diff
@@ -1636,7 +1641,7 @@ def check_requirements(self, show_warning=False):
                     if not pkg in installed_packages:
                         missing.append(pkg)
 
-                if missing and install_requirements:
+                if missing and install_requirements and not show_warning_only:
```
Comment (bridadan) [inline mbed/mbed.py]: Oh hang on, so now the `show_warning_only` is being used to select if things should be auto installed? That doesn't sound right.

File: mbed/mbed.py (hunk 1 above)
Comment (screamerbg) [reply mbed/mbed.py]: And I already explained that not reminding that we're skipping all requirements check (which is NOT RECOMMENDED) should be clearly visible. Read my comment.

File: mbed/mbed.py (hunk 6 above)
Comment (screamerbg) [reply mbed/mbed.py]: show_warning_only = do not auto-install

Comment (theotherjimmy) [reply mbed/mbed.py]: @screamerbg Perhaps it could use a different name then?

Comment (screamerbg) [reply mbed/mbed.py]: @theotherjimmy Proposal?

Comment (theotherjimmy) [reply mbed/mbed.py]: `require_install`?and invert the logic so that True raises an error and False warns.

File: mbed/mbed.py (hunk 5 above)
Comment (screamerbg) [reply mbed/mbed.py]: To clarify the behavior:
* Mbed CLI should attempt to auto-install requirements when:
   * Codebased is being imported (`mbed import`, `mbed add`, `mbed deploy`, `mbed new`)
   * Codebased is being updated/checked out (`mbed update`)
   * Certificates are being installed or changed (`mbed dm`)
* Mbed CLI should only check and NOT try to auto-install requirements when:
   * Code is being compiled, tested or exported
   * Any Mbed CLI operation that doesn't imply network connectivity

As thoroughly described here https://github.com/ARMmbed/mbed-cli/issues/756. Also as mentioned in my first comment to this PR.

> Related to #795, #782 and #756

File: mbed/mbed.py (hunk 4 above)
Comment (screamerbg) [reply mbed/mbed.py]: To clarify the behavior:
* Mbed CLI should attempt to auto-install requirements when:
   * Codebased is being imported (`mbed import`, `mbed add`, `mbed deploy`, `mbed new`)
   * Codebased is being updated/checked out (`mbed update`)
   * Certificates are being installed or changed (`mbed dm`)
* Mbed CLI should only check and NOT try to auto-install requirements when:
   * Code is being compiled, tested or exported
   * Any Mbed CLI operation that doesn't imply network connectivity

As thoroughly described here https://github.com/ARMmbed/mbed-cli/issues/756. Also as mentioned in my first comment to this PR.

> Related to #795, #782 and #756

**Comment by screamerbg** [comment]
To clarify the behavior:
* Mbed CLI should attempt to auto-install requirements when:
   * Codebased is being imported (`mbed import`, `mbed add`, `mbed deploy`, `mbed new`)
   * Codebased is being updated/checked out (`mbed update`)
   * Certificates are being installed or changed (`mbed dm`)
* Mbed CLI should only check and NOT try to auto-install requirements when:
   * Code is being compiled, tested or exported
   * Any Mbed CLI operation that doesn't imply network connectivity

As thoroughly described here https://github.com/ARMmbed/mbed-cli/issues/756. Also as mentioned in my first comment to this PR.

> Related to #795, #782 and #756

In addition to this, the behavior above can be overridden using the `NO_REQUIREMENTS` config flag, which is a **severe** deviation from the standard and established behavior of Mbed CLI (NOT RECOMMENDED) and it should be made clearly visible at all times.

**Comment by bridadan** [comment]
@screamerbg Thanks for the clarifications, I hadn't seen that issue before, that helps give the context.

I still would prefer to have the two flags separate in `check_requirements`. If you like I could add that in a follow up PR?

**Comment by screamerbg** [comment]
Hey man, no worries! I could have made that clearer, but I had to rush to meetings 😄
